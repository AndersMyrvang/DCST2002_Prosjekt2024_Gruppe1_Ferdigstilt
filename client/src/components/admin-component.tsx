import React, { useState, useEffect } from 'react';
import { User } from '../types';

const AdminPanel = () => {
  const [users, setUsers] = useState<User[]>([]);

  const fetchUsers = () => {
    fetch('/api/users')
      .then((response) => response.json())
      .then((data) => setUsers(data));
  };

  useEffect(() => {
    fetchUsers();

    const interval = setInterval(fetchUsers, 60000);

    return () => clearInterval(interval);
  }, []);

  // Sjekker om en bruker er online basert på last_login og last_logout
  const isUserOnline = (user: User) => {
    if (
      user.last_login &&
      (!user.last_logout || new Date(user.last_login) > new Date(user.last_logout))
    ) {
      return true; 
    }
    return false;
  };

  // Funksjon for å endre admin-status
  const toggleAdminStatus = (userId: number, isAdmin: boolean) => {
    fetch(`/api/users/${userId}/set-admin`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ isAdmin: !isAdmin }),
    })
      .then((response) => {
        if (response.ok) {
          setUsers(
            users.map((user) => (user.user_id === userId ? { ...user, is_admin: !isAdmin } : user)),
          );
        } else {
          console.error('Feil ved oppdatering av admin-status');
        }
      })
      .catch((error) => console.error('Feil ved toggleAdminStatus:', error));
  };

  return (
    <div className="admin-panel">
      <table>
        <thead>
          <tr>
            <th>User ID</th>
            <th>Username</th>
            <th>Email</th>
            <th>Member Since</th>
            <th>Online Status</th>
            <th>Admin</th>
          </tr>
        </thead>
        <tbody>
          {users.map((user) => (
            <tr key={user.user_id}>
              <td>{user.user_id}</td>
              <td>{user.username}</td>
              <td>{user.email}</td>
              <td>{new Date(user.first_login).toLocaleDateString()}</td>
              <td>
                <span className={isUserOnline(user) ? 'online-dot' : 'offline-dot'}></span>
              </td>
              <td>
                <input
                  type="checkbox"
                  checked={user.is_admin}
                  onChange={() => toggleAdminStatus(user.user_id!, user.is_admin)}
                />
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default AdminPanel;
