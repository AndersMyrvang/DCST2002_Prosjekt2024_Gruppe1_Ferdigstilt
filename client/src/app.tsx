import React, { useEffect, useState } from 'react';
import { HashRouter as Router, Route } from 'react-router-dom';
import NavBar from './components/navBar-components';
import { PlayerDetails } from './components/player-components';
import { TeamDetails } from './components/team-components';
import { LeagueDetails } from './components/league-components';

// Dummy Home component as placeholder
function Home() {
  return <div>Welcome Home</div>;
}

function App() {
  interface User {
    displayName: string;
  }

  const [user, setUser] = useState<User | null>(null);

  // Fetching user info from /api/current_user when the app loads
  useEffect(() => {
    fetch('/api/current_user')
      .then((res) => res.json())
      .then((data) => {
        setUser(data);
      });
  }, []);

  return (
    <Router>
      {/* Define routes for the page */}
      <Route exact path="/" component={Home} />
      <Route path="/players/:name" component={PlayerDetails} />
      <Route path="/teams/:name" component={TeamDetails} />
      <Route path="/leagues/:name" component={LeagueDetails} />
    </Router>
  );
}

export default App;
