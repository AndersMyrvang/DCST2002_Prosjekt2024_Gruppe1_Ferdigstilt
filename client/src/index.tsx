import * as React from 'react';
import { createRoot } from 'react-dom/client';
import { Component } from 'react-simplified';
import { HashRouter, Route } from 'react-router-dom';
import { NavBar, Card, Alert } from './widgets';
import {
  PlayerList,
  PlayerDetails,
  PlayerEdit,
  PlayerListAll,
  PlayerNew,
} from './components/player-components';
import {
  TeamList,
  TeamDetails,
  TeamEdit,
  TeamListAll,
  TeamNew,
} from './components/team-components';
import { TagList, TagAdd } from './components/tag-component';
import {
  LeagueList,
  LeagueDetails,
  LeagueEdit,
  LeagueListAll,
  LeagueNew,
} from './components/league-components';
import Hamburger from './components/navBar-components';
import Profile from './components/profile-component';

class Menu extends Component {
  render() {
    return <Hamburger user={null} />;
  }
}

class Home extends Component {
  render() {
    return (
      <Card title="Welcome">
        <PlayerList />
        <TeamList />
        <LeagueList />
      </Card>
    );
  }
}

let root = document.getElementById('root');
if (root)
  createRoot(root).render(
    <HashRouter>
      <div>
        <Alert />
        <Menu />
        <Route exact path="/" component={Home} />
        <Route exact path="/menu" component={Hamburger} />

        <Route exact path="/page" component={PlayerList} />
        <Route exact path="/players" component={PlayerListAll} />
        <Route exact path="/player/:id" component={PlayerDetails} />
        <Route exact path="/player/:id/edit" component={PlayerEdit} />
        <Route exact path="/player_id/new" component={PlayerNew} />

        <Route exact path="/page" component={TeamList} />
        <Route exact path="/teams" component={TeamListAll} />
        <Route exact path="/team/:id/" component={TeamDetails} />
        <Route exact path="/team/edit/:id/" component={TeamEdit} />
        <Route exact path="/team_id/new" component={TeamNew} />

        <Route exact path="/page" component={LeagueList} />
        <Route exact path="/leagues" component={LeagueListAll} />
        <Route exact path="/league/:id" component={LeagueDetails} />
        <Route exact path="/league/:id/edit" component={LeagueEdit} />
        <Route exact path="/league_id/new" component={LeagueNew} />

        <Route exact path="/tags" component={TagList} />
        <Route exact path="/tags/new" component={TagAdd} />
        <Route exact path="/profile" component={Profile} />
      </div>
    </HashRouter>,
  );
