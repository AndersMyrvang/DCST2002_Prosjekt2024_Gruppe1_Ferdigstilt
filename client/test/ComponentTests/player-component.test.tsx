import React from 'react';
import { render, screen, fireEvent, waitFor, act, logRoles } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { PlayerListAll, PlayerList, PlayerDetails, PlayerEdit, PlayerNew } from '../../src/components/player-components';
import playerService from '../../src/services/player-service';
import userService from '../../src/services/user-service';
import tagService from '../../src/services/tag-service';
import { Alert } from '../../src/widgets';
import CommentSection from '../../src/components/comment-components';
import commentService from '../../src/services/comment-service';

jest.mock('../../src/services/player-service', () => ({
  listAllPlayers: jest.fn(),
  listSetAmountOfPlayers: jest.fn(),
  getPlayer: jest.fn(),
  updatePlayer: jest.fn(),
  deletePlayer: jest.fn(),
  addPlayer: jest.fn(),
  getCountries: jest.fn(),
  getTeams: jest.fn(),
  incrementViewCount: jest.fn().mockResolvedValue(undefined),
  getCreator: jest.fn().mockResolvedValue({ username: "Creator", page_id: 1, created_at: '2021-05-01', view_count: 1 }),
  getPageHistory: jest.fn().mockResolvedValue({ username: "Creator", page_id: 1, created_at: '2021-05-01', view_count: 1 }),
  getLeague: jest.fn(),
  getRevisedName: jest.fn().mockResolvedValue({ name: 'Player 1' }),
}));

jest.mock('../../src/services/user-service', () => ({
  getUserById: jest.fn(),
}));

jest.mock('../../src/services/tag-service', () => ({
  getTagsForPage: jest.fn(),
  addTagToPage: jest.fn(),
  removeTagFromPage: jest.fn(),
  getTags: jest.fn().mockResolvedValue([]),
}));

jest.mock('../../src/services/comment-service', () => ({
  listComments: jest.fn().mockResolvedValue([]),
  addComment: jest.fn().mockResolvedValue(1),
}));

jest.mock('../../src/services/league-service');
jest.mock('../../src/services/team-service');
jest.mock('../../src/services/revision-service', () => ({
  getRevisions: jest.fn().mockResolvedValue([]),
}));

jest.mock('../../src/widgets', () => ({
  ...jest.requireActual('../../src/widgets'),
  Alert: {
    danger: jest.fn((message) => {
      console.error(message); 
    }),
  },
}));


beforeEach(() => {
  jest.clearAllMocks();
  (playerService.listAllPlayers as jest.Mock).mockResolvedValue([
    {
      id: 1,
      name: 'Player 1',
      birth_date: '2000-01-01',
      height: 180,
      country_name: 'Country 1',
      team_name: 'Team 1',
    },
    {
      id: 2,
      name: 'Player 1', 
      birth_date: '1995-05-05',
      height: 175,
      country_name: 'Country 2',
      team_name: 'Team 2',
    },
  ]);
  jest.spyOn(Storage.prototype, 'getItem').mockImplementation((key) => {
    if (key === 'isLoggedIn') return 'true';
    if (key === 'user') return JSON.stringify({ user_id: 1, username: 'TestUser' });
    return null;
  });
});

const renderWithRouter = (component: JSX.Element) => render(<Router>{component}</Router>);

describe('PlayerListAll Component', () => {
  test('renders all players', async () => {
    renderWithRouter(<PlayerListAll />);

    const heading = await screen.findByText('All Players');
    expect(heading).toBeInTheDocument();

    const players = await screen.findAllByText(/Player 1/);
    expect(players).toHaveLength(2);
  });


  test('does not show "New Player" button if user is not logged in', () => {
    jest.spyOn(Storage.prototype, 'getItem').mockImplementation(() => null);
    renderWithRouter(<PlayerListAll />);
    const button = screen.queryByText('New player');
    expect(button).toBeNull();
  });




});

describe('PlayerDetails Component', () => {
  const mockPlayer = {
    id: 1,
    name: 'Player 1',
    birth_date: '2000-01-01',
    height: 180,
    country_name: 'Country 1',
    team_name: 'Team 1',
    page_id: 1,
    content: 'Player bio content',
  };
  const mockTags = [{ tag_id: 1, tag_name: 'Tag 1' }];

  const renderWithRouter = (component: JSX.Element) => render(<Router>{component}</Router>);

  beforeEach(() => {
    jest.clearAllMocks();
    (playerService.getPlayer as jest.Mock).mockResolvedValue(mockPlayer);
    (tagService.getTagsForPage as jest.Mock).mockResolvedValue(mockTags);
  });


  test('renders player details', async () => {
    renderWithRouter(<PlayerDetails match={{ params: { id: 1 } }} />);

    const playerHeadings = await screen.findAllByText(/Player 1/);
    expect(playerHeadings).toHaveLength(3);

    const mainHeading = await screen.findByRole('heading', { level: 2 });
    expect(mainHeading).toHaveTextContent('Player 1');

    const cardTitle = await screen.findByText((content, element) => {
      return element?.tagName === 'H5' && content === 'Player 1';
    });
    expect(cardTitle).toBeInTheDocument();

    const tagsHeading = await screen.findByText(/Tags for Player 1/);
    expect(tagsHeading).toBeInTheDocument();
  });

  test('renders associated tags correctly', async () => {
    mockPlayer.page_id = 1;

    const mockTags = [
      { tag_id: 1, tag_name: 'Tag A' },
      { tag_id: 2, tag_name: 'Tag B' },
    ];

    (playerService.getPlayer as jest.Mock).mockResolvedValue(mockPlayer);
    (tagService.getTagsForPage as jest.Mock).mockResolvedValue(mockTags);

    await act(async () => {
      render(
        <PlayerDetails match={{ params: { id: 1 } }} />
      );
    });

    const tagElements = await screen.findAllByRole('listitem');
    expect(tagElements).toHaveLength(2);
    expect(tagElements[0]).toHaveTextContent('Tag A');
    expect(tagElements[1]).toHaveTextContent('Tag B');
  });
});


test('displays fallback message when no tags are available', async () => {
  (playerService.getPlayer as jest.Mock).mockResolvedValue({
    id: 1,
    name: 'Player 1',
    page_id: 1,
    birth_date: '2000-01-01',
    height: 180,
    country_name: 'Country 1',
    team_name: 'Team 1',
    content: 'Player bio content',
  });

  (tagService.getTagsForPage as jest.Mock).mockResolvedValue([]); 

  renderWithRouter(<PlayerDetails match={{ params: { id: 1 } }} />);

  
  const fallbackMessage = await screen.findByText('No tags available for this player.');
  expect(fallbackMessage).toBeInTheDocument();
});

describe('PlayerEdit Component', () => {
  const mockPlayer = {
    id: 1,
    name: 'Player 1',
    birth_date: '2000-01-01',
    height: 180,
    country_id: 1,
    team_id: 1,
    picture_url: '',
    page_id: 1,
    content: 'Player content',
  };

  beforeEach(() => {
    jest.clearAllMocks();
    (playerService.getPlayer as jest.Mock).mockResolvedValue(mockPlayer);
    (playerService.getCountries as jest.Mock).mockResolvedValue([
      { country_id: 1, country_name: 'Country 1' },
    ]);
    (playerService.getTeams as jest.Mock).mockResolvedValue([{ id: 1, name: 'Team 1' }]);
    (tagService.getTagsForPage as jest.Mock).mockResolvedValue([{ tag_id: 1, tag_name: 'Tag 1' }]);
  });

  const renderWithRouter = (component: JSX.Element) => render(<Router>{component}</Router>);

  test('prefills form fields with current player data', async () => {
    renderWithRouter(<PlayerEdit match={{ params: { id: 1 } }} />);
    const nameInput = await screen.findByDisplayValue('Player 1');
    expect(nameInput).toBeInTheDocument();
  });

  test('navigates to the player with id', async () => {
    renderWithRouter(<PlayerEdit match={{ params: { id: 1 } }} />);

  
    const button = await screen.findByTestId('playerBackID');
    fireEvent.click(button);

    expect(window.location.href).toContain('/player/1');
  });

  test('adds a tag to a player', async () => {
    const mockPlayerTags = [{ tag_id: 2, tag_name: 'Tag B' }];
    const Harry = [{
      id: 1,
      name: 'Harry Kane',
      birth_date: new Date('1993-07-28'),
      height: 188,
      country_id: 1,
      country_name: 'England',
      team_name: 'Tottenham Hotspur',
      team_id: 101,
      picture_url: 'https://example.com/player1.png',
      page_id: 201,
      content: 'A prolific striker from England.',
      league_id: 1,
      league_name: 'Premier League',
    }];
    const mockTeam = {
      id: 1,
      name: 'Team A',
      country_id: 1,
      country_name: 'Country A',
      coach: 'Coach A',
      league_id: 1,
      league_name: 'League A',
      emblem_image_url: '',
      content: '',
      page_id: 1,
    };

    const mockTags = [
      { tag_id: 1, tag_name: 'Tag A' },
      { tag_id: 2, tag_name: 'Tag B' },
    ];

    (playerService.getPlayer as jest.Mock).mockResolvedValue(mockPlayer);
    (tagService.getTags as jest.Mock).mockResolvedValue(mockTags);
    (tagService.addTagToPage as jest.Mock).mockResolvedValue(undefined);
    (tagService.getTagsForPage as jest.Mock).mockResolvedValue(mockTags);

    await act(async () => {
      render(
        <PlayerEdit match={{ params: { id: 1 } }} />
      );
    });

    fireEvent.change(screen.getByDisplayValue('Select tag to add'), { target: { value: '1' } });
    fireEvent.click(screen.getByText(/Add Tag/i));

    await waitFor(() => {
      expect(tagService.addTagToPage).toHaveBeenCalledWith(mockPlayer.page_id, 1);
    });

    expect(screen.getAllByText('Tag A').length).toBe(2);
  });

  test('removes a tag from a team', async () => {
    const mockTags = [
      { tag_id: 1, tag_name: 'Tag A' },
      { tag_id: 2, tag_name: 'Tag B' },
    ];

    const mockPlayer = {
      id: 1,
      name: 'Player 1',
      birth_date: new Date('1993-07-28'),
      height: 188,
      country_id: 1,
      country_name: 'England',
      team_name: 'Tottenham Hotspur',
      team_id: 101,
      picture_url: 'https://example.com/player1.png',
      page_id: 201,
      content: 'A prolific striker from England.',
      league_id: 1,
      league_name: 'Premier League',
    };

    (playerService.getPlayer as jest.Mock).mockResolvedValueOnce(mockPlayer);
    (tagService.getTagsForPage as jest.Mock).mockResolvedValueOnce(mockTags);
    (tagService.removeTagFromPage as jest.Mock).mockResolvedValueOnce(undefined);

    await act(async () => {
      render(
        <PlayerEdit match={{ params: { id: 1 } }} />
      );
    });

    expect(screen.getAllByText('Tag A')).toHaveLength(2);
    expect(screen.getAllByText('Tag B')).toHaveLength(2);

    const removeButton = screen.getByLabelText('Remove Tag A');
    fireEvent.click(removeButton);

    await waitFor(() => {
      expect(tagService.removeTagFromPage).toHaveBeenCalledWith(201, 1);
    });
  });

  describe('PlayerNew Component', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      (playerService.getCountries as jest.Mock).mockResolvedValue([
        { country_id: 1, country_name: 'Country 1' },
      ]);
      (playerService.getTeams as jest.Mock).mockResolvedValue([{ id: 1, name: 'Team 1' }]);
    });

    const renderWithRouter = (component: JSX.Element) => render(<Router>{component}</Router>);

    test('renders new player form', () => {
      renderWithRouter(<PlayerNew />);
      expect(screen.getByText('New Player')).toBeInTheDocument();
    });

    test('submits new player form successfully', async () => {
      (playerService.addPlayer as jest.Mock).mockResolvedValue(123);

      renderWithRouter(<PlayerNew />);

      fireEvent.change(screen.getByLabelText(/name/i), { target: { value: 'New Player' } });
      fireEvent.change(screen.getByLabelText(/height/i), { target: { value: '190' } });
      fireEvent.change(screen.getByLabelText(/birth date/i), { target: { value: '2000-01-01' } });
      fireEvent.click(screen.getByText(/save/i));

      await waitFor(() => expect(playerService.addPlayer).toHaveBeenCalled());
      expect(playerService.addPlayer).toHaveBeenCalledWith(
        'New Player',
        '2000-01-01',
        190,
        expect.anything(),
        expect.anything(), 
        expect.anything(), 
        expect.anything() 
      );
    });

    test('navigates to all players on button click', async () => {
      renderWithRouter(<PlayerNew />);

      const button = await screen.findByTestId('playerAllBack');
      fireEvent.click(button);

      expect(window.location.href).toContain('/players');
    });

  });

  describe('PlayerList Component', () => {
    beforeEach(() => {
      (playerService.listSetAmountOfPlayers as jest.Mock).mockResolvedValue([
        { id: 1, name: 'Player 1', height: 180, country_name: 'Country 1', team_name: 'Team 1' },
        { id: 2, name: 'Player 2', height: 175, country_name: 'Country 2', team_name: 'Team 2' },
      ]);
      document.body.innerHTML = '<div id="alerts"></div>';

    });

    const renderWithRouter = (component: JSX.Element) => render(<Router>{component}</Router>);

    test('renders a list of players', async () => {
      renderWithRouter(<PlayerList />);
      const players = await screen.findAllByText(/Player/i);
      expect(players).toHaveLength(2);
    });

    test('renders limited number of players', async () => {
      (playerService.listSetAmountOfPlayers as jest.Mock).mockResolvedValue([
        { id: 1, name: 'Player 1', height: 180, country_name: 'Country 1', team_name: 'Team 1' },
      ]);

      renderWithRouter(<PlayerList />);

      const playerLinks = await screen.findAllByRole('link', { name: /Player/i });
      expect(playerLinks).toHaveLength(1);
    });

    test('navigates to all players on button click', async () => {
      renderWithRouter(<PlayerList />);

      const button = await screen.findByTestId('playerAllButton');
      fireEvent.click(button);

      expect(window.location.href).toContain('/players');
    });

    test('displays an error message when fetching teams fails', async () => {
      (playerService.listAllPlayers as jest.Mock).mockRejectedValue(new Error('Network Error'));

      render(
        <PlayerListAll />
      );

      expect(playerService.listAllPlayers).toHaveBeenCalled();
    });


  });

  describe('Comments Component', () => {
    test('renders comments', async () => {
      const mockPlayer = {
        id: 1,
        name: 'Player 1',
        birth_date: new Date('1993-07-28'),
        height: 188,
        country_id: 1,
        country_name: 'England',
        team_name: 'Tottenham Hotspur',
        team_id: 101,
        picture_url: 'https://example.com/player1.png',
        page_id: 201,
        content: 'A prolific striker from England.',
        league_id: 1,
        league_name: 'Premier League',
      };

      const mockComments = [
        { comment_id: 1, user_id: 1, content: 'Great team!', username: 'User A' },
        { comment_id: 2, user_id: 2, content: 'Amazing players!', username: 'User B' },
      ];

      (playerService.getPlayer as jest.Mock).mockResolvedValue(mockPlayer);
      (commentService.listComments as jest.Mock).mockResolvedValue(mockComments);

      await act(() => {
        render(
          <PlayerDetails match={{ params: { id: 1 } }} />
        );
      });

      await waitFor(() => {
        expect(screen.getByText('Great team!')).toBeInTheDocument();
        expect(screen.getByText('Amazing players!')).toBeInTheDocument();
      });
    });

    test('increments view count on mount', async () => {
      const mockPlayer = {
        id: 1,
        name: 'Player 1',
        birth_date: new Date('1993-07-28'),
        height: 188,
        country_id: 1,
        country_name: 'England',
        team_name: 'Tottenham Hotspur',
        team_id: 101,
        picture_url: 'https://example.com/player1.png',
        page_id: 201,
        content: 'A prolific striker from England.',
        league_id: 1,
        league_name: 'Premier League',
      };

      (playerService.getPlayer as jest.Mock).mockResolvedValue(mockPlayer);
      (playerService.getPageHistory as jest.Mock).mockResolvedValue([]);
      (playerService.getLeague as jest.Mock).mockResolvedValue({ league_id: 1, league_name: 'Premier League' });
      (playerService.getCreator as jest.Mock).mockResolvedValue({ user_id: 1, username: 'Creator' });
      (tagService.getTags as jest.Mock).mockResolvedValue([]);
      (playerService.getRevisedName as jest.Mock).mockResolvedValue({ name: 'Player 1' });

      (tagService.getTags as jest.Mock).mockResolvedValue([
        { tag_id: 1, tag_name: 'Tag A' },
        { tag_id: 2, tag_name: 'Tag B' },
        { tag_id: 3, tag_name: 'Tag C' },
      ]);

      (tagService.getTagsForPage as jest.Mock).mockResolvedValue([
        { tag_id: 1, tag_name: 'Tag A' },
        { tag_id: 2, tag_name: 'Tag B' },
        { tag_id: 3, tag_name: 'Tag C' },
      ]);

      (tagService.getTagsForPage as jest.Mock).mockResolvedValue([
        { tag_id: 2, tag_name: 'Tag B' },
      ]);

      await act(async () => {
        render(
          <PlayerDetails match={{ params: { id: 1 } }} />
        );
      });

      await waitFor(() => {
        expect(playerService.incrementViewCount).toHaveBeenCalledWith(mockPlayer.page_id);
      });
    });

    test('Displays page history', async () => {
      const mockPlayer = {
        id: 1,
        name: 'Player 1',
        birth_date: new Date('1993-07-28'),
        height: 188,
        country_id: 1,
        country_name: 'England',
        team_name: 'Tottenham Hotspur',
        team_id: 101,
        picture_url: 'https://example.com/player1.png',
        page_id: 201,
        content: 'A prolific striker from England.',
        league_id: 1,
        league_name: 'Premier League',
      };

      const mockRevision = [
        { revision_id: 1, user_id: 1, username: 'User A', revised_at: '2021-05-01', revised_name: 'Player 1' },
      ];


      (playerService.getPlayer as jest.Mock).mockResolvedValue(mockPlayer);
      (playerService.getPageHistory as jest.Mock).mockResolvedValue(({ username: "Creator", page_id: 1, created_at: '2021-05-01', view_count: 1 }));

      await act(async () => {
        render(
          <PlayerDetails match={{ params: { id: 1 } }} />
        );
      });

      await waitFor(() => {
        expect(screen.getByText('User A')).toBeInTheDocument();
        expect(screen.getAllByText('Player 1').length).toBeGreaterThan(1);
      });
    });

    test('handles invalid page history gracefully', async () => {
      jest.clearAllMocks();
      (playerService.getPageHistory as jest.Mock).mockResolvedValue({ username: null, created_at: null, created_by: null, view_count: 1 });
      (playerService.getCreator as jest.Mock).mockResolvedValue({ username: null, created_at: null });
      renderWithRouter(<PlayerDetails match={{ params: { id: 1 } }} />);
      await waitFor(() => {
        expect(screen.getByText(/Unknown/i)).toBeInTheDocument();
        expect(screen.getByText(/1.1.1970/i)).toBeInTheDocument();
      });
    });

    /* test('adds a new comment', async () => {
      jest.clearAllMocks();
      (playerService.getPageHistory as jest.Mock).mockResolvedValue({ username: null, created_at: null });
    
  
      renderWithRouter(<PlayerDetails match={{ params: { id: 1 } }} />);
  
      await waitFor( () => {
        fireEvent.change(screen.getByLabelText(/Comment input/i), { target: { value: 'Nice player!' } });
        fireEvent.click(screen.getByText(/Send/i));
        expect(screen.getByText(/Nice player!/i)).toBeInTheDocument();
      }) 
    });
   */
    test('displays error if fetching players fails', async () => {
      (playerService.listAllPlayers as jest.Mock).mockRejectedValue(new Error('Network error'));
      renderWithRouter(<PlayerListAll />);
      await waitFor(() => expect(screen.getByText('Error fetching players')));
    });
  });

  describe('PlayerComponent snapshots', () => {

    test('matches PlayerDetails snapshot', () => {
      const { asFragment } = renderWithRouter(<PlayerDetails match={{ params: { id: 1 } }} />);
      expect(asFragment()).toMatchSnapshot();
    });

  });
});
