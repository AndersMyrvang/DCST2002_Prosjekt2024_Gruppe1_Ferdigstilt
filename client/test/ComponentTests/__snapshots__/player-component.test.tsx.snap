// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerEdit Component PlayerComponent snapshots matches PlayerDetails snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-light"
    type="button"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
      />
    </svg>
     Back to all players
  </button>
  <div
    class="card card1"
  >
    <div
      class="card-body card-body1"
    >
      <h5
        class="card-title card-title1"
      />
      <div
        class="card-text card-text1"
      >
        <h2 />
        <div
          class="row"
        >
          <div
            class="col-7"
          >
            <div
              class="float-start"
            >
              <div
                class="row"
              >
                <div
                  class="card card4 "
                >
                  <div
                    class="card-body card-body4 "
                  >
                    <h5
                      class="card-title card-title4"
                    />
                    <div
                      class="card-content card-content4"
                    >
                      <div />
                      <div
                        class="row"
                      >
                        <div
                          class="col"
                        >
                          <div
                            class="float-start"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="col-3"
          >
            <div
              class="float-start"
            >
              <div
                class="row"
              >
                <div
                  class="card card2 "
                >
                  <div
                    class="card-body card-body2 "
                  >
                    <h5
                      class="card-title card-title2"
                    />
                    <div
                      class="card-content card-content2"
                    >
                      <div
                        class="col-11"
                      >
                        <div
                          class="float-start"
                        >
                          <div
                            class="card card4 "
                          >
                            <div
                              class="card-body card-body4 "
                            >
                              <h5
                                class="card-title card-title4"
                              />
                              <div
                                class="card-content card-content4"
                              >
                                <p>
                                  <b>
                                    Birth date:
                                  </b>
                                   19.11.2024
                                </p>
                                <p>
                                  <b>
                                    Height:
                                  </b>
                                   0 cm
                                </p>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="card card1"
        >
          <div
            class="card-body card-body1"
          >
            <h5
              class="card-title card-title1"
            >
              Page History
            </h5>
            <div
              class="card-text card-text1"
            >
              <div
                style="margin-left: 60px; display: flex; width: 1080px; justify-content: space-evenly; align-items: center;"
              >
                <div
                  style="display: flex; justify-content: space-evenly; width: 100%;"
                >
                  <p
                    style="flex: 1; text-align: left; margin-right: 20px;"
                  >
                    Laget av: Unknown, 19.11.2024
                  </p>
                  <p
                    style="flex: 1; text-align: center; margin-right: 20px;"
                  >
                    Sist oppdatert av: Denne siden har ikke blitt revidert enda19.11.2024
                  </p>
                  <p
                    style="flex: 1; text-align: right;"
                  >
                    View count: 0
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    style="margin-top: 20px;"
  >
    <h3>
      Tags for 
    </h3>
    <p>
      No tags available for this player.
    </p>
  </div>
  <div
    class="card card1"
  >
    <div
      class="card-body card-body1"
    >
      <h5
        class="card-title card-title1"
      >
        Comments
      </h5>
      <div
        class="card-text card-text1"
      />
    </div>
  </div>
</DocumentFragment>
`;
